### DISCUSSION TITLE
Logger class in a scheduled job

### DISCUSSION THREAD

**Discussion by pnicosia80** [oqtane/oqtane.framework#3567] [Q&A] [answered]
Good evening,
I need to instantiate the Logger class in a scheduled job to write information to the oqtane event log.
How can I do?
It's not enough for me to write in the scheduled job log.

Thank you

**Comment by sbwalker** [comment, marked as answer]
@pnicosia80 Oqtane is multi-tenant... so the challenge in a scheduled job is that there is no context... which means that the framework is unable to determine the tenant/site on its own. This is a problem because the LogManager service (and most other Oqtane services) requires this context in order to log information to the database for the appropriate site. So in this scenario you will need to explicitly provide the context:

ie.

```
// get services
var tenantManager = provider.GetRequiredService<ITenantManager>();
var siteRepository = provider.GetRequiredService<ISiteRepository>();
var logManager = provider.GetRequiredService<ILogManager>();

// get tenant (set by HostedServiceBase)
var tenantId = tenantManager.GetAlias().TenantId;

foreach (var site in siteRepository.GetSites().ToList())
{
    // explicitly set tenant and site
    var alias = new Alias { TenantId = tenantId, SiteId = site.SiteId };
    tenantManager.SetAlias(alias);

    // log information
    logManager.Log(LogLevel.Information, this, LogFunction.Other, "Logging Information For Alias {Alias}", alias);
}
```

**Reply by pnicosia80** [reply]
It worked with some modifications, i.e. I had to find the tenantId with GetTenant, because GetAlias was null (why??).

if(_tenantId < 0)
{
    var currAlias = _tenantManager.GetAlias();
    var currTenant = _tenantManager.GetTenant();

    // get tenant (set by HostedServiceBase)
    if (currAlias != null)
        _tenantId = _tenantManager.GetAlias().TenantId;  //GetAlias() is null

    if (currTenant != null)
        _tenantId = _tenantManager.GetTenant().TenantId;
}

Also I had to assign a name to the alias, otherwise it was null and the _logManager.Log went into exception

   // explicitly set tenant and site
   var alias = new Alias { TenantId = _tenantId, SiteId = site.SiteId, Name = this.Name };
   _tenantManager.SetAlias(alias);
   
   
The exception was in the Alias class in the ".Contains" because the Name was null, maybe a check would be needed if the Name is null:

 public string Path
 {
     get
     {
         if (Name.Contains("/"))
         {
             return Name.Substring(Name.IndexOf("/") + 1);
         }
         else
         {
             return "";
         }
     }
 }

Thanks

**Reply by sbwalker** [reply]
@pnicosia80 I should have clarified that the code I provided was simply an example of how to solve the problem you described - it had not actually been tested to determine if it was functional (there is no example in the core framework which uses logging in a scheduled job). Regardless, I am glad you were able to get it to work - and I appreciate you pointing out a few issues which can be addressed.

In regards to your question of why GetAlias is returning null in this scenario... I can see the logic which is causing it - but the logic has existed for 2 years and seems to work fine in all other scenarios. So I will need to investigate further.

**Reply by sbwalker** [reply]
@pnicosia80 #3575 fixes the issue with the Alias.Path property. After investigating the behavior of GetAlias I have determined that it has a validation check to ensure that the Alias object is fully hydrated and is not a "mock" object. This is because a lot of other code relies on the fact that GetAlias returns a valid object. So in your scenario GetTenant() is the correct method to use because you want to retrieve the TenantId - not the Alias. 

That being said, if you need to identify a specific tenant/site instance in a scheduled job then you do need to construct and set an Alias object. Currently the best way to do this is to use the AliasRepository and retrieve an Alias based on TenantId and SiteId - and then call SetAlias(). To simplify this in the future I have now added a new method to TenantManager so that it is possible to SetAlias using a TenantId and SiteId (which populates an Alias object automatically).

**Reply by sbwalker** [reply]
So in 5.0.1, the following pattern should work:

```
// get services
var tenantManager = provider.GetRequiredService<ITenantManager>();
var siteRepository = provider.GetRequiredService<ISiteRepository>();
var logManager = provider.GetRequiredService<ILogManager>();

// get tenant (set by HostedServiceBase)
var tenantId = tenantManager.GetTenant().TenantId;

foreach (var site in siteRepository.GetSites().ToList())
{
    // explicitly set alias
    tenantManager.SetAlias(tenantId, site.SiteId);

    // log information
    logManager.Log(LogLevel.Information, this, LogFunction.Other, "Logging Information For Site {SiteId} and Tenant {TenantId}", site.SiteId, tenantId);
}
```

**Reply by pnicosia80** [reply]
Thanks for the release, I confirm that everything worked in my local environment. Congrats on the efficiency!

Have a good evening
